### ISSUE TITLE
NT-Xent Loss Implementation Breakdown

### ISSUE THREAD

**Issue by KarahanS** [KarahanS/Self-Supervised-Learning-for-Medical-Image-Analysis#7]
We discussed whether the current InfoNCE (or NT-Xent) loss implementation is correct or not (or the same as the one provided [here](https://github.com/sthalles/SimCLR/blob/master/simclr.py)). Here is a breakdown of the `loss` function (assuming a batch size of 256 and an output dimension of 128 (for projection head)):
```python
  x, _ = batch
  x = torch.cat(x, dim=0)  # convert the list of images to tensor     x = (256, 3, 28, 28)

  # x = 512 images
  # for each image, we apply two augmentations:
  # x[0] - positive pair #1
  # x[0 + batch_size] - positive pair #2

  # Apply base encoder and projection head to images to get embedded representations
  z = self.forward(x)

  # z = (512, output_dim)    
  # z[:, None, :] = (512, 1, 128)  (just reshape)
  # z[None, :, :] = (1, 512, 128)  (just reshape)
```
We want to calculate the following loss function. As a normalization, we would like to use all the cosine similarities except the similarities between the same elements (cosine similarity to itself)
![image](https://github.com/KarahanS/Self-Supervised-Learning-for-Medical-Image-Analysis/assets/44376034/789f5d3e-a130-48df-bf0b-d074ffc41a80)

Apply cosine similarity between each entry - producing a matrix of 512x512. Diagonal entries correspond to the cosine similarity of each entry with itself.
```python
cos_sim = F.cosine_similarity(z[:, None, :], z[None, :, :], dim=-1)
```
Mask out diagonal entries.
```python
self_mask = torch.eye(cos_sim.shape[0], dtype=torch.bool, device=cos_sim.device)
        cos_sim.masked_fill_(
            self_mask, -9e15
        )  # set cosine similarity to itself to -9e15 (nearly 0)
```

Roll the mask matrix by half.
```python
pos_mask = self_mask.roll(shifts=cos_sim.shape[0] // 2, dims=0)
```
An example input-output of this procedure:
```
in = tensor([[1,  0,  0,  0],
        [ 0, 1, 0, 0],
        [ 0, 0, 1, 0],
        [ 0, 0, 0, 1]])
out = tensor([[ 0, 0, 1, 0],
        [0, 0, 0, 1],
        [ 1,  0,  0,  0],
        [ 0,  1,  0,  0]])
```
We want to calculate the cosine similarity between each positive pair, and normalize it with the all cosine similarities. Let's apply the rolling operation on our simple example. In this case, we deal with 4 images (feature embeddings). As a result of the rolling operation, we get a mask that would use the followings as the positive example:
- cosine similarity between 0th and 2nd indices
- cosine similarity between 1st and 3rd indices
- cosine similarity between 0th and 3rd indices
- cosine similarity between 1st and 2nd indices
This makes sense, as the `x` would be a list as follows:
[1st augmentation on image 1,  1st augmentation on image 2, 2nd augmentation on image 1, 2nd augmentation on image 2]

Then we calculate the mean loss:
```python
cos_sim /= self.hparams.temperature
loss = (-cos_sim[pos_mask] + torch.logsumexp(cos_sim, dim=-1)).mean()
```

## Conclusion
The current implementation is **exactly** the same as the one provided in [SimCLR paper](https://arxiv.org/pdf/2002.05709). 

## Metrics

```python
  # Get ranking position of positive example
  comb_sim = torch.cat(
      [cos_sim[pos_mask][:, None], cos_sim.masked_fill(pos_mask, -9e15)],
      # First position positive example
      dim=-1,
  )

  sim_argsort = comb_sim.argsort(dim=-1, descending=True).argmin(dim=-1)

  # Logging ranking metrics
  self.log(mode + "_acc_top1", (sim_argsort == 0).float().mean())
  self.log(mode + "_acc_top5", (sim_argsort < 5).float().mean())
  self.log(mode + "_acc_mean_pos", 1 + sim_argsort.float().mean())
 ```
 
 We had some difficulty understanding these metrics. Basically, these metrics assume that it is 2-view SimCLR (as in the original paper). It tries to find for each "first augmentations", the other "second augmentation". If the most similar representation to a first augmentation is the second augmentation, then we increase the top-1 accuracy. If the second augmentation is within the top 5, then we increase the top-5 accuracy. ```_acc_mean_pos``` represents the position of the second augmentation within the ranking of similarities. The lower it gets the better it is.
